File: blind-75/arrays-and-hashing/valid-anagram/go/valid_anagram_test.go
package validanagram

import (
	"strings"
	"testing"
)

func TestValidAnagram(t *testing.T) {
	tests := []struct {
		name 	string
		s 		string
		t 		string
		want 	bool
	}{
		{"Valid Anagram with RACE", "racecar", "carrace", true},
		{"Invalid Anagram with JAR adn JAM", "jar", "jam", false},
		{"Invalid Anagram with x", "x", "x", true},
		{"Valid anagram", "anagram", "nagaram", true},
		{"Valid with repeats", "aacc", "ccaa", true},
		{"Same string", "abc", "abc", true},
		{"Single char equal", "x", "x", true},
		{"Empty vs non-empty", "", "a", false},
		{"Prefix but shorter", "ab", "a", false},
		{"Longer s", "abc", "ab", false},
		{"Classic invalid", "rat", "car", false},
		{"Same letters different counts", "aab", "abb", false},
		{"One char differs", "jar", "jam", false},
		{"All same vs mixed", "aaaa", "aaab", false},
		{"Both empty", "", "", true},
		{"Single char differs", "x", "y", false},
		{"Long repeated valid", "aaaaaaaaaa", "aaaaaaaaaa", true},
		{"Longer t", "ab", "abc", false},               // early-return other direction — you only have longer-s
		{"Distinct permuted", "abcdef", "fedcba", true}, // no-repeats path, your valids all repeat
		{"Long invalid at end", strings.Repeat("a", 50000)+"b", strings.Repeat("a", 50000)+"c", false}, // late mismatch + max-len stress — you only havelong-valid
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := ValidAnagram(tt.s, tt.t); got != tt.want {
				t.Errorf("ValidAnagram(%v, %v) = %v want %v", tt.s, tt.t, got, tt.want)
			}
		})
	}
}

File: blind-75/arrays-and-hashing/valid-anagram/typescript/valid-anagram.test.ts
import { expect, test } from "bun:test";
import { validAnagram } from "./valid-anagram";

test("Valid Anagram with RACE", () => {
  expect(validAnagram("racecar", "carrace")).toBe(true);
});

test("Invalid Anagram with JAR adn JAM", () => {
  expect(validAnagram("jar", "jam")).toBe(false);
});

test("Invalid Anagram with x", () => {
  expect(validAnagram("x", "x")).toBe(true);
});

test("Valid anagram", () => {
  expect(validAnagram("anagram", "nagaram")).toBe(true);
});

test("Valid with repeats", () => {
  expect(validAnagram("aacc", "ccaa")).toBe(true);
});

test("Same string", () => {
  expect(validAnagram("abc", "abc")).toBe(true);
});

test("Single char equal", () => {
  expect(validAnagram("x", "x")).toBe(true);
});

test("Empty vs non-empty", () => {
  expect(validAnagram("", "a")).toBe(false);
});

test("Prefix but shorter", () => {
  expect(validAnagram("ab", "a")).toBe(false);
});

test("Longer s", () => {
  expect(validAnagram("abc", "ab")).toBe(false);
});

test("Classic invalid", () => {
  expect(validAnagram("rat", "car")).toBe(false);
});

test("Same letters different counts", () => {
  expect(validAnagram("aab", "abb")).toBe(false);
});

test("One char differs", () => {
  expect(validAnagram("jar", "jam")).toBe(false);
});

test("All same vs mixed", () => {
  expect(validAnagram("aaaa", "aaab")).toBe(false);
});

test("Both empty", () => {
  expect(validAnagram("", "")).toBe(true);
});

test("Single char differs", () => {
  expect(validAnagram("x", "y")).toBe(false);
});

test("Long repeated valid", () => {
  expect(validAnagram("aaaaaaaaaa", "aaaaaaaaaa")).toBe(true);
});

test("Longer t", () => {
  expect(validAnagram("ab", "abc")).toBe(false);
});

test("Distinct permuted", () => {
  expect(validAnagram("abcdef", "fedcba")).toBe(true);
});

test("Long invalid at end", () => {
  expect(validAnagram("a".repeat(50000) + "b", "a".repeat(50000) + "c")).toBe(false);
});
